File: 2022/OpenCV4 tutorial/aruco/aruco.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/aruco.hpp>
#include <iterator>

using namespace cv;


int main(int argc, const char* argv[])
{
    Mat img = imread("./aruco.jpg");
    std::vector< int > markerIds;
    std::vector< std::vector<Point2f> > markerCorners;
    aruco::detectMarkers(img, aruco::getPredefinedDictionary(cv::aruco::DICT_5X5_100), markerCorners, markerIds);
    for (int i = 0; i < markerIds.size(); i++) {
        Point2f topLeft = markerCorners.at(i).at(0);
        Point2f topRight = markerCorners.at(i).at(1);
        Point2f bottomRight = markerCorners.at(i).at(2);
        Point2f bottomLeft = markerCorners.at(i).at(3);
        line(img, topLeft, topRight, Scalar(0, 255, 0), 5);
        line(img, topRight, bottomRight, Scalar(0, 255, 0), 5);
        line(img, bottomRight, bottomLeft, Scalar(0, 255, 0), 5);
        line(img, bottomLeft, topLeft, Scalar(0, 255, 0), 5);
        Point2f c((topLeft.x + bottomRight.x) / 2.0,
            (topLeft.y + bottomRight.y) / 2.0);

        circle(img, c, 4, Scalar(0, 0, 255), -1);


        putText(img, std::to_string(markerIds.at(i)),
                Point2f(topLeft.x, topLeft.y - 15),
                   FONT_HERSHEY_SIMPLEX,
                   2, Scalar(0, 0, 255), 5);

    }


    namedWindow("aruco", WINDOW_NORMAL);
    imshow("aruco", img);
    waitKey();
    return 0;
}

File: 2022/OpenCV4 tutorial/aruco/aruco.cs
using System;
using System.Collections.Generic;
using System.IO;
using OpenCvSharp;

namespace Stitching
{
    class Program
    {

        static void Main(string[] args)
        {
            DirectoryInfo rootDir = Directory.GetParent(Environment.CurrentDirectory);
            string root = rootDir.Parent.Parent.FullName;
            System.IO.Directory.SetCurrentDirectory(root);

            //IList<IList<Point>> pts = new List<IList<Point>>(); // 存放多个多边形点的集合
            //pts.Add(new List<Point>()); // 添加第一条曲线
            Mat outImg = Cv2.ImRead(@".\aruco.jpg");
            Point2f[][] corners, rejectedImgPoints;
            int[] ids;
            OpenCvSharp.Aruco.CvAruco.DetectMarkers(outImg, OpenCvSharp.Aruco.CvAruco.GetPredefinedDictionary(OpenCvSharp.Aruco.PredefinedDictionaryName.Dict5X5_100), out corners, out ids, OpenCvSharp.Aruco.DetectorParameters.Create(), out rejectedImgPoints);
            for (int i = 0; i < ids.Length; i++)
            {
                //Console.WriteLine("{0}", ids[i]);
                Point topLeft = new Point(corners[i][0].X, corners[i][0].Y);
                Point topRight = new Point(corners[i][1].X, corners[i][1].Y);
                Point bottomRight = new Point(corners[i][2].X, corners[i][2].Y);
                Point bottomLeft = new Point(corners[i][3].X, corners[i][3].Y);

                Cv2.Line(outImg, topLeft, topRight, new Scalar(0, 255, 0), 5);
                Cv2.Line(outImg, topRight, bottomRight, new Scalar(0, 255, 0), 5);
                Cv2.Line(outImg, bottomRight, bottomLeft, new Scalar(0, 255, 0), 5);
                Cv2.Line(outImg, bottomLeft, topLeft, new Scalar(0, 255, 0), 5);
                Point c = new Point((topLeft.X + bottomRight.X) / 2.0, (topLeft.Y + bottomRight.Y) / 2.0);

                Cv2.Circle(outImg, c, 4, new Scalar(0, 0, 255), -1);
                Cv2.PutText(outImg, ids[i].ToString(),
                new Point(topLeft.X, topLeft.Y - 15),
                   HersheyFonts.HersheySimplex,
                   2, new Scalar(0, 0, 255), 5);
            }
        

            Cv2.ImShow("out", outImg);
            Cv2.WaitKey();

        }
    }
}
